### PR TITLE
JavaScript: Fix indentation for binary expressions with trailing comments

### LINKED ISSUE #8451: Comment breaks pipeline and logic operator indention
**Prettier pr-8435**
[Playground link](https://deploy-preview-8435--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDAOlABFgPgPi1S0xwKJwHpKtIBbOhGUvQjbVozTYgMl5IcANtADmAGTGSoEqXJnTZCsVn4sRyzUu2Ld80aoFZqtCAybr9erdZ0q1wq05vPR3KMVy5LL33b943o5+tqH6gVQ09IywPv7xYVqBcYkJriAANCAQAA4wAJbQAM7IoKgATuUQAO4AChUIJSioQtWoAJ4lWQBG5ahgANZwMADKOf35Msgw5QCucFlwdN1wACara+KoMrOoonAAYhDldKgwBVPNszAQmSAAFjB0QgDq9-nwReNgcCONH-kAG4fdrIcBFLogSZFODlGC1PqiU7IABmLRhWQAVkUAB4AIT6g2GI1QjHEkzgqPRCxA2JxI0moiEcAAirMIPAqUIMSBxuUYeUwd1UCshHccuVJjAXvlVjB7sgABwABiyEogMJefRyYIlcAFgMpWQAjuz4Ajck00EUALRQOBrNZ3cpwU35F0IvbIpBo7k0mF0fLTOb+xnMtkcyk+6lZGAimVyhVIABMsb6+SEjIAwmZvSB9QBWO6zGEAFRFTV9PMB8wAklANrARmBJXkAIINkYwdrMrkwgC+-aAA) (pipeline has not released yet, playground is from a PR)
```sh
--parser babel
```

**Input:**
```jsx
a
  |> a 
  |> a  // comment
  |> a
  |> a

a && 
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong &&  // comment
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong

a ||
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong ||  // comment
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong
```

**Output:**
```jsx
a
|> a
|> a  // comment
  |> a
  |> a;

a &&
longLongLongLongLongLongLongLongLong &&
longLongLongLongLongLongLongLongLong && // comment
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong;

a ||
longLongLongLongLongLongLongLongLong ||
longLongLongLongLongLongLongLongLong || // comment
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong;

```

**Expected behavior:**

Should aligned

<details>
<summary>Assign work as expected</summary>

**Prettier pr-8435**
[Playground link](https://deploy-preview-8435--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuc0DOMAEZMF5MCGAOlJpgD4B8hmJZVNmA9E9hALbsIx0XXGk+hEiUhQM2PDQBk02oIA20AOYAZFeqhqNOrZu16VmWbyWHzBy-uu7lxuc1aRO3U7ZsWPVoycXv-ngHKIlBiEjj4BBTkboFx3vHRsQkpXl7RZCxsLrDJaUH5Wkl+8YWpKiAANCAQAA4wAJboyKAEAE5tEADuAArtCGjIIAQKXQQAnoPVAEZtBGAA1nAwAMq18w1ayDBtAK5w1XDs03AAJqdnqgRauwTKcABiEG3sBDCNWygEuzAQVSAACxg7AUAHUAQ14Gh1mA4CsBpCGgA3SHjIZgNBTECbNBwNowHpzZSvZAAMxGuOqACs0AAPABCc0WyxWBC4qk2cDJFIOIBptJWm2UCjgAEVdhB4NyFJSQOs2ri2kNpgQTgp-rU2psYKCGqcYADkAAOAAM1U1EFxoLmtSGmrgiqRXOqAEcJfBCXVBl80ABaKBwM5nf5tOBuhqhwl3ElIcky3m49gNbZ7BNCkXiyVc2M86owVW6-WGpAAJjzcwaCiFAGEODGQA6AKz-Xa4gAqqu9cdlSP2AEkoBdYCswFr6gBBQcrGDjEXS3EAXwXQA)
```sh
--parser babel
```

**Input:**
```jsx
const c = a
  |> a 
  |> a  // comment
  |> a
  |> a

const c = a && 
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong &&  // comment
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong

const c = a ||
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong ||  // comment
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong
```

**Output:**
```jsx
const c =
  a
  |> a
  |> a  // comment
  |> a
  |> a;

const c =
  a &&
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong && // comment
  longLongLongLongLongLongLongLongLong &&
  longLongLongLongLongLongLongLongLong;

const c =
  a ||
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong || // comment
  longLongLongLongLongLongLongLongLong ||
  longLongLongLongLongLongLongLongLong;

```
</details>

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Fixes #8451
This PR seems to work fine but the doc is transformed so it maybe a dirty way...

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Seems dirty, can it done in `printBinaryishExpressions`?

**Comment by sosukesuzuki** [comment]
okay, I'll try it later.

**Comment by sosukesuzuki** [comment]
done

**Comment by fisker** [comment]
I mean can we print the first part differently in that function?

**Comment by thorn0** [comment]
Performance-wise, it was better before the change. `printBinaryishExpressions` is recursive, and we don't need to do this flattening on each recursion level.

**Comment by thorn0** [comment]
**Prettier pr-8476**
[Playground link](https://deploy-preview-8476--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAjAlgMwgJygYR3SgHM4cBZCAZwAs4ATABXIAcAbOADwGUBDAO4NyACgA6UAASS+1ANYAVHBABuCPgCE46gCJxWCBtQBKEAXyjUANBOkl26ALbU4AcXYBPVrWoAJPgBefDgMAHIQEKzUCuhwMXA2UjJqODiMGnww9DjUGhBcyhCOGgCuJCQeoSU47BIAlJISAD4AfJJyEGC0fKj4EbGufI5wAPJQI6gQnACqUAJwfGrNbfZOLu5ePv5BIeGR0bHxANwSEhjYeITEZJQ09MxsnLyCwjjiSbKKympQmtq-egMUCMpnMlkSdgczjcnm8fkCwTCESi8XiEOS5DSDAyWXIuXyhWKZQqVRq9UkAHoKZJII5hrBlu1Ot1ev03ENRuNJjM5gsllBWpJVtCNnDtoi9ijDrEjiArCBIjB0NBqMhQMFlAImMEEKqUHx2OYPKr5agcHwwHI4DAeKwLddkDAcCUEiA4I5UIxhAwADIWEglPhkABiuEcmSVpGQID4JRgEDlIFoMEc7AA6rR0PBqHawHAeLqs+gVFmPNGwNQTSBiC4cDAmOaSOHkJgDS55QArahcDTmy3W-jDH3EOAttuuru8a6cACKJQg8DH7HbIDtOXI0dQPTg7ETrCIsDT6AYWWQAA4AAzy-c0OBp82saP7uC1tSJgCO8-gDf20dkAFooG0b1EzST90DSBsg2bJBW2XV0XEcdBHWdBDpzgOcF1HWDx3lGAeiPE9aGQAAmPDzXQBxSD6Ok+GjF8AFZExKFwFB6PU4JXFQXQASWBBAbTAIhWBgABBYEeBgDxOCXFwAF85KAA)
```sh
--parser babel
```

**Input:**
```jsx
bifornCringerMoshedPerplexSawder(
  askTrovenaBeenaDependsRowans,
  glimseGlyphsHazardNoopsTieTie,
  averredBathersBoxroomBuggyNurl
) 
|> kochabCooieGameOnOboleUnweave
|> glimseGlyphsHazardNoopsTieTie;

bifornCringerMoshedPerplexSawder(
  askTrovenaBeenaDependsRowans,
  glimseGlyphsHazardNoopsTieTie,
  averredBathersBoxroomBuggyNurl
) // comment
|> kochabCooieGameOnOboleUnweave
|> glimseGlyphsHazardNoopsTieTie;
```

**Output:**
```jsx
bifornCringerMoshedPerplexSawder(
  askTrovenaBeenaDependsRowans,
  glimseGlyphsHazardNoopsTieTie,
  averredBathersBoxroomBuggyNurl
)
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

bifornCringerMoshedPerplexSawder(
    askTrovenaBeenaDependsRowans,
    glimseGlyphsHazardNoopsTieTie,
    averredBathersBoxroomBuggyNurl
  )  // comment
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

```

**Comment by sosukesuzuki** [comment]
@fisker 
> I mean can we print the first part differently in that function?

I tried but I was not sure how to print it differently...

@thorn0
> Performance-wise, it was better before the change. printBinaryishExpressions is recursive, and we don't need to do this flattening on each recursion level.

the flattening is actually done only when the parent node type is `ExpressionStatement` so I think this is no problem for performance.

**Comment by thorn0** [comment]
@sosukesuzuki I'd try to do flattening where `printComments` is called. `parts` is an array everywhere, assigning it the result of the `printComments` call, which is a `Doc { type: "concat" }`, doesn't look right. See also https://github.com/prettier/prettier/issues/8451#issuecomment-637567706

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -4963,6 +4963,25 @@ function printBinaryishExpressions(
...
+  // For alignment expression statements with trailing comments like:
+  //
+  // foo +
+  //   bar + // comment
+  //   baz;
+  const hasTrailingCommentInBinaryish = (node) => {
+    if (isBinaryish(node) && !node.comments) {
+      return hasTrailingCommentInBinaryish(node.left);
+    }
+    return isBinaryish(node) && hasTrailingComment(node);
```
Comment (thorn0) [inline src/language-js/printer-estree.js:4975]: **Prettier pr-8476**
[Playground link](https://deploy-preview-8476--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAjAlgMwgJygYR3SgHM4cBZCAZwAs4ATABXIAcAbOADwGUBDAO4NyACgA6UAASS+1ANYAVHBABuCPgCE46gCJxWCBtQBKEAXyjUANBOkl26ALbU4AcXYBPVrWoAJPgBefDgMAHIQEKzUCuhwMXASAJQSAD4AfJLYEKkZchBgtHyo+BGxrnyOcADyUFWoEJwAqlACcHxqOZL2Ti7uXj7+QSHhkdGx8QDcEhIY2HiExGSUNPTMbJy8gsI44lIy8kqq6lq6+oYmZhbWtl0Ozm6e3n6BwWERUfHxSZ1ZkgD0f0kkEclVgnTyBSKJQgZQq1Vq9SaLTaHSg6VuPQe-WeQzeo0+sQmICsIEiMHQ0GoyFAwWUAiYwQQVJQfHY5g8VJJqBwfDAcjgMB4rF5i2QMBwAFc4CS4I5UIxhAwADIWEgSvhkABiuEcfBg5NIyBAfAlMAgxJAtBgjnYAHVaOh4NRhWA4DwmY70CpHR4jWBqJyQMQXDgYEweSRdchMKyXCSAFbULgaHl8gX8SpK4hwaOx6UgRO8RacACKEog8Fz7DjIGFOBDRtQRTg7AtrCIsFt6AYMFoyAAHAAGEntmhwW081hG9twENqC0AR3L8HDoyNsgAtFBtIqLTg4Ev0Pvwxqo0gY9X8y5HOgxZKr8W4GWKznz3mSTAil2e32kAAmD8eXQBxSBKEE+CNWcAFYLQlFwFCKZkLxrFQpQASSgYRYB4MAiFYGAAEFMJ4GAPE4KsXAAX0ooA)
```sh
--parser babel
```

**Input:**
```jsx
bifornCringerMoshedPerplexSawder(
  askTrovenaBeenaDependsRowans,
  glimseGlyphsHazardNoopsTieTie
)
|> foo
|> kochabCooieGameOnOboleUnweave
|> glimseGlyphsHazardNoopsTieTie;

bifornCringerMoshedPerplexSawder(
  askTrovenaBeenaDependsRowans,
  glimseGlyphsHazardNoopsTieTie
)
|> foo // comment
|> kochabCooieGameOnOboleUnweave
|> glimseGlyphsHazardNoopsTieTie;
```

**Output:**
```jsx
bifornCringerMoshedPerplexSawder(
  askTrovenaBeenaDependsRowans,
  glimseGlyphsHazardNoopsTieTie
)
  |> foo
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

bifornCringerMoshedPerplexSawder(
    askTrovenaBeenaDependsRowans,
    glimseGlyphsHazardNoopsTieTie
  )
  |> foo  // comment
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

```

**Comment by sosukesuzuki** [comment]
@thorn0 How do you think about adding an option named `shouldReturnParts` to `comments.printComments`.  See https://github.com/prettier/prettier/pull/8476/commits/4ce5bf36f837f5c05e7857db25950bd9d36ad33a

**Comment by thorn0** [comment]
Looks complicated and I'm not sure it's okay to circumvent the `prependCursorPlaceholder` calls (need to investigate, I never used `formatWithCursor`).

I have another idea. There is a function called `normalizeParts` in `language-html/utils`. It isn't specific to HTML, so let's extract it to `document/doc-utils` and call it on the result of our `printComments` calls.

**Comment by fisker** [comment]
This one seem also fixes #5183

File: src/language-js/printer-estree.js [hunk 2]
```diff
@@ -1159,9 +1159,12 @@ function printPathNoParens(path, options, print, args) {
...
       ]);
 
-      // We group here when the callee is itself a call expression.
-      // See `isLongCurriedCallExpression` for more info.
-      if (isCallOrOptionalCallExpression(n.callee)) {
+      if (
+        // We group here when the callee is itself a call expression.
+        // See `isLongCurriedCallExpression` for more info.
+        isCallOrOptionalCallExpression(n.callee) ||
+        isBinaryish(path.getParentNode())
```
Comment (thorn0) [inline src/language-js/printer-estree.js:1166]: No, we can't tackle it this way:

**Prettier pr-8476**
[Playground link](https://deploy-preview-8476--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAjAlgMwgJygYR3SgHM4cBZCAZwAs4ATABXIAcAbOADwGUBDAO4NyAbQA6UAASS+1ANYAVHBABuCPgCE46gCJxWCBtQBKEAXyjVJAagnSS7dAFtqcAOLsAnq1rUAEnwAXnw4DAByEBCs1ArocLFwEgC6dpIAPgB8ktgQ0gD0eZKQTk4IMKmZknIQYLR8qPiRcW58pQDyUG2oEJwAqlACcHxqFVkOzq4e3r4BwaERUTFxCQDcEiAANCBRMOjQ1MigIcoCTCEIByh87OaeB1uoOHxgcnAwPKzPxCTIMDgArnAtnAnKhGMIGAAZCwkf58MgAMVwTj4MF2pGQID4-xgEE2IFoMCc7AA6rR0PBqJ8wHAeBcKegVBTPJiwNR7iBiK4cDAmE8SCjkJhrq4tgArahcDRPF5vfilSHEOBCkVAkAS3jfTgARX+EHgKvYopAnxw3MxqHqcHY+NYRFgJPQDBgtGQAA4AAxbO00OAkp6sTF2uDctT4gCOevgfMWmNkAFooNoIficHBI+g03z4YKkMKjWrXE50L8AYWtXBdfrlXnVVsYPVHc7XUgAEz1p7oRykRolPiYkMAVnx-1cCnql3zxpUgIAklBhLAeGAiKwYABBBc8GCeTiG1wAXwPQA)
```sh
--parser babel
```

**Input:**
```jsx
bifornCringerMoshedPerplexSawder[
  askTrovenaBeenaDependsRowans +
  glimseGlyphsHazardNoopsTieTie
]
  |> foo  // comment
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

```

**Output:**
```jsx
bifornCringerMoshedPerplexSawder[
    askTrovenaBeenaDependsRowans + glimseGlyphsHazardNoopsTieTie
  ]
  |> foo  // comment
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

```

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js:1166]: I know, I'm fixing it.:+1:

Comment (sosukesuzuki) [reply src/language-js/printer-estree.js]: done https://github.com/prettier/prettier/pull/8476/commits/d4e5e6cc79be51ed417d603c48b25a8a9c3a67a5

File: src/document/doc-utils.js [hunk 3]
```diff
@@ -219,6 +219,37 @@ function stripTrailingHardline(doc, withInnerParts = false) {
...
+    ) {
+      newParts.push(newParts.pop() + part);
+      continue;
+    }
+
+    newParts.push(part);
+  }
+
+  return newParts;
+}
```
Comment (evilebottnawi) [inline src/document/doc-utils.js:251]: We export this file to plugins, so I think we should documented it

Comment (evilebottnawi) [reply src/document/doc-utils.js:251]: https://github.com/prettier/prettier/blob/cedb1a2c2539fcc5813bc4aa4169401d5ea2d7cd/src/index.js#L50
https://github.com/prettier/prettier/blob/cedb1a2c2539fcc5813bc4aa4169401d5ea2d7cd/src/document/index.js

Comment (evilebottnawi) [reply src/document/doc-utils.js:251]: I was wrong, we don't have documentation about utils

Comment (thorn0) [reply src/document/doc-utils.js:251]: This is definitely out of scope for this PR. The entire `prettier.doc` namespace isn't documented.

**Comment by thorn0** [comment]
@sosukesuzuki I think we came close to fixing https://github.com/prettier/prettier/issues/1210.

**Comment by sosukesuzuki** [comment]
@fisker 
> This one seem also fixes #5183

I think #5183 is the other problem. This doesn't related to comments.

**Comment by fisker** [comment]
Block comment seem ugly

**Prettier pr-8476**
[Playground link](https://deploy-preview-8476--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBtARgSwGYQE5QDCemUA5nHgLIQDOAFnACYAKlADgDZwAeAygEMA7k0oBdADpQAPgD4ABLgjypcxRGUB6AFTzIAW30IY87ZtUKl8nXoiHjp8zIUBrCGHoD0hDZjgBxASMAeShg9AhuAFUoITgBADc4C3kyTkx9WgDOAE92eloACQEALwE8JgA5DXZaABU-BrgAbhAAGhAIdhhMaFpkUHK8CCEWcoR+lAFOIQEc-o70PAEwFzgYPnYV0jJkGDwAVzgOuH10ZlEmABkBcgOBCgAxfH0BGB7yZBABA5gIdpA9Bg+k4AHV6Jh4LQtmA4HwJpDMAlITkvmBaAsQKQsngYCxlmRXshsNMsh0AFa0HgAIWWq3WgiMV1IcGJpOOIEp-B23AAigcIPA2ZwySAtngcV90F44JwAewSLBQZgmDB6MgABwABg6CrocFBy3YXwVcBxSQBAEcBfB8V1Jt9aABaKBwZjMAF4ODWzBe-EPIlIEkijlZfSYYWi2g8uD8wWsoPsjowLzK1XqpAAJmTy0w6XIPkMAi+ZoArACDlk6l5JsHRQkjgBJKCiWB8MAkboAQRbfBgOW4kbgAF9h0A)
```sh
--parser babel
--no-semi
```

**Input:**
```jsx
[bifornCringerMoshedPerplexSawder]
|> foo 
|> foo /* comment */
|> foo /* comment */
|> kochabCooieGameOnOboleUnweave
|> glimseGlyphsHazardNoopsTieTie;
```

**Output:**
```jsx
;[bifornCringerMoshedPerplexSawder]
  |> foo
  |> foo |>
  /* comment */
  foo |>
  /* comment */
  kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie

```

**Prettier 2.0.5**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBtARgSwGYQE5QDCemUA5nHgLIQDOAFnACYAKlADgDZwAeAygEMA7k0oBdADpQAPgD4ABLgjypcxRGUB6AFTzIAW30IY87ZtUKl8nXoiHjp8zIUBrCGHoD0hDZjgBxASMAeShg9AhuAFUoITgBADc4C3kyTkx9WgDOAE92eloACQEALwE8JgA5DXZaABU-BrgAbhAAGhAIdhhMaFpkUHK8CCEWcoR+lAFOIQEc-o70PAEwFzgYPnYV0jJkGDwAVzgO+hh9TgB1ekx4Wi2wOD4Jm8wEm5zkcFoFkFIsvBgLGWZH0AmQ2GmWQ6ACtaDwAELLVbrQRGAAypDg4MhxxAsP4O24AEUDhB4NjOFCQFs8P9PugvHBOO1qSRYBdMEwYPRkAAOAAMHXYwyyF2W7E+wrg-ySLIAjqT4ECupMQAJaABaKBwZjMFl4OAKzAGoECEFgpAQym4rL6TB7Q42wlwElkrGWnEdGBeDlcnlIABMXuWmHS5B8hgtIGlAFYWQcsnUvJMrVSEkcAJJQUSwPhgEjdACC2b4MBy3ApWQAvlWgA)
```sh
--parser babel
```

**Input:**
```jsx
[bifornCringerMoshedPerplexSawder]
|> foo 
|> foo /* comment */
|> foo /* comment */
|> kochabCooieGameOnOboleUnweave
|> glimseGlyphsHazardNoopsTieTie;
```

**Output:**
```jsx
[bifornCringerMoshedPerplexSawder]
|> foo
|> foo /* comment */
|> foo /* comment */
  |> kochabCooieGameOnOboleUnweave
  |> glimseGlyphsHazardNoopsTieTie;

```

**Comment by thorn0** [comment]
@fisker This is an oversight from #7787.

**Pull request merged**